py/desispec/io/spectra.py

```diff
@@ -212,11 +233,20 @@ def read_spectra(infile, single=False):
         raise IOError("{} is not a file".format(infile))
 
     t0 = time.time()
-    hdus = fitsio.FITS(infile, mode='r')
+    hdus = fitsio.FITS(infile, mode="r")
     nhdu = len(hdus)
+    ahdus = fits.open(infile)
+
+    if targetids is not None:
+        targetids = np.atleast_1d(targetids)
+        file_targetids = hdus["FIBERMAP"].read(columns="TARGETID")
+        rows = np.where(np.isin(file_targetids, targetids))[0]
```

Comment:
As implemented, this will return a Spectrum with the targets ordered by how they appear in the input file, not ordered by how they appear in the `targetids` list.  At minimum that "feature" should be documented so that the user doesn't assume that `output.fibermap['TARGETID'] == targetids`.

I would find it slightly more natural to return them in the order they appear in the input `targetids` list.  I don't know of an elegant way to do that, but a brute force way would be:
```python
rows = np.concatenate([np.where(file_targetids == tid)[0] for tid in targetids])
```
Note that that still supports the possibility that some `targetids` aren't in the `file_targetids` (good).

Reply:
Quick aside: note that `fitsio>1.1.9` (https://github.com/esheldon/fitsio/blob/master/CHANGES.md#version-119) is the first version to support returning rows in the order that you requested, so the brute-force sorting (I wish there was a faster way than looping...) would need to happen at the end and all the HDUs would need to be reindexed in the general case.

Reply:
if useful - and if I m correct - `desitarget.geomask.match_to` could be an alternative solution to avoid looping; like
```
rows = match_to(file_targetids, targetids)
```
though this assumes that both `file_targetids` and `targetids` don t have any duplicates (for `file_targetids`, I suspect that s always true when one deals with spectra files; one could think harder about early stuck fibers).
this method is super-fast, but that may not be relevant here as the spectra files won t contain zillions of rows; still, in case, you could still some piece of code from there.

and also it may be useful to make here simple checks on `targetids` and return a warning/error; for instance, what if the provided `targetids` have duplicates? or values not present in the spectra file?

Reply:
Thanks for the warning about fitsio versions and the tip about `match_to`.  Unfortunately we do need to support targetids being repeated in the input files since that occurs for uncoadded spectra, and `match_to` doens't support that case.  I ended up implementing this as a "re-order at the end only if needed" including supporting duplicated and missing TARGETIDs, with lots of unit tests because it is a bit tricky.